### ISSUE TITLE
const enum error clarification

### ISSUE THREAD

**Issue by Skillz4Killz** [denoland/deno#7605]
This error appears after I switched to Deno 1.4.1 and fixed all the import type errors.

---

![image](https://user-images.githubusercontent.com/23035000/93725703-1f229580-fb7f-11ea-858b-ee5d04457f43.png)

```ts
Skillz@Home Discordeno % deno run --allow-net --allow-read --unstable debug.ts
error: Uncaught ReferenceError: RequestMethod is not defined
    return runMethod(RequestMethod.Get, url, body);
                     ^
    at Object.get (requestManager.ts:124:22)
    at createClient (client.ts:44:41)
    at debug.ts:10:1
```

```ts
  get: async (url: string, body?: unknown) => {
    return runMethod(RequestMethod.Get, url, body);
  },
```

![image](https://user-images.githubusercontent.com/23035000/93725685-0a460200-fb7f-11ea-8a1a-1bdd9c6d77ec.png)

![image](https://user-images.githubusercontent.com/23035000/93725689-0d40f280-fb7f-11ea-9785-464b0b9a73f5.png)

Also of note VSC shows 0 errors
![image](https://user-images.githubusercontent.com/23035000/93725697-1762f100-fb7f-11ea-9f6f-73bb47393324.png)

Alternative Fix: (to just removing `const`)
![image](https://user-images.githubusercontent.com/23035000/93725706-2649a380-fb7f-11ea-9876-2f698986350d.png)

Reason(By Kit <3): const enum are not supportable in a --no-check situation, and so in Deno 1.4.0+ we switched on the isolatedModules flag with --unstable to help folks ensure their code was "no check" safe. 

Can we get a better error message for this please. Confused me like crazy! Although im always confused so... hahaha

**Comment by nayeemrmn** [comment]
This is https://github.com/microsoft/TypeScript/issues/40499. The error message isn't the problem, it's a TS bug.

**Comment by kitsonk** [comment]
Ooops, yeah, forgot about that.  Tracking issue then for that.

**Comment by tbillington** [comment]
Looks like the mentioned issue https://github.com/microsoft/TypeScript/issues/40499 was closed by this PR https://github.com/microsoft/TypeScript/pull/41933.

I'm still getting this issue in Deno 1.7.1.

**Comment by lucacasonato** [comment]
@tbillington That has not yet made it into the TS release in Deno.

**Comment by tbillington** [comment]
Makes sense, thank you :)

**Comment by lucacasonato** [comment]
Actually this will also require some work in SWC before we can ship it. It seems like the change is that const enums are now treated like regular enums when `isolatedModules` is enabled (`isolatedModules: true` now implies `preserveConstEnums: true`).

cc @kdy1

**Comment by tbillington** [comment]
So the "const" part is/will be effectively erased?

**Comment by kdy1** [comment]
Yes, const part is ignored.
It's same as babel, but well.. Do you think it worth implementing?

I'm going to implement one along with my type checker, but it would take some time.

**Comment by nayeemrmn** [comment]
What changes are required in swc? The bug never occurred under `--no-check`.

**Comment by kdy1** [comment]
I'm not sure, but it would be a huge change.
TypeScript works by loading all dependency modules implicitly, while swc and babel works on each file.

It requires makeing swc load dependencies (maybe via a trait) and analyze them.

**Comment by nayeemrmn** [comment]
But that's to implement non-`isolatedModules` rules. If we assume `isolatedModule` rules on both the imported and importing modules (an assumption which holds for Deno), it just needs to transpile as a regular `enum` and emit the appropriate `import`/`export` statements on each end.

Rather, the complication exists for Node ecosystem usages where the transpiler can't assume that the prebuilt third party lib emitted `const enum`s as regular `enum`s. We have it easy.

**Comment by kdy1** [comment]
Does an user write import statement for const enums when isolatedModules is true?

**Comment by nayeemrmn** [comment]
You have to write import statements anyway, the question is if those imports are still there in the emit. That's now the expected behaviour of both tsc (with `isolatedModules`) and Deno, and it was already the behaviour of swc.

**Comment by kitsonk** [comment]
I agree with @nayeemrmn there is nothing for swc to do here for Deno.

**Comment by nayeemrmn** [comment]
Fixed by #9341.

**Comment by jflatow** [comment]
I don't understand the resolution here -- upstream dependencies may have `const enum` and there's nothing we can do downstream -- the safest option would be to allow a downstream package to specify that it wants to copy the const enums while still requiring a cast between the types. But it is currently impossible to do this without copy/pasting the whole enum, which is much more fragile.
